daemon.go

```diff
@@ -59,12 +60,27 @@ func (daemon menderDaemon) StopDaemon() {
 
 func (daemon *menderDaemon) Run() error {
 	// figure out the state
-	switch daemon.GetState() {
-	case MenderFreshInstall:
-		//do nothing
-	case MenderRunningWithFreshUpdate:
-		daemon.CommitUpdate()
+	for {
+		switch daemon.TransitionState() {
+		case MenderStateRunningWithFreshUpdate:
+			daemon.CommitUpdate()
+		case MenderStateWaitForUpdate:
+			err := daemon.waitForUpdate()
+			if err != nil {
+				return err
```

Comment:
Well the problem is that fetch update, appply update and commit update should be separate states/state action. Having that you could easily decide about the next transition. Right now we can only guess based on error returned.